### ISSUE TITLE
Metal peformance issue: Sentence Transformer using Candle gets aborted by the system

### ISSUE THREAD

**Issue by aorticweb** [huggingface/candle#1596]
When using `sentence-transformers/all-MiniLM-L6-v2` on m1 PRO mac (32GB ram) and attempting to generate embeddings for 100 sentences. the program get's aborted by the system after GPU usage spikes to 100.

### How to reproduce:
- place the following code snipet (below) in `candle-examples/examples/bert-metal/main.rs`
- cargo run --example bert-metal  --features metal

```
#[cfg(feature = "mkl")]
extern crate intel_mkl_src;

#[cfg(feature = "accelerate")]
extern crate accelerate_src;
use candle_transformers::models::bert::{BertModel, Config, DTYPE};

use anyhow::{Error as E, Result};
use candle::Tensor;
use candle_nn::VarBuilder;
use hf_hub::{api::sync::Api, Repo, RepoType};
use tokenizers::{PaddingParams, Tokenizer};

fn build_model_and_tokenizer() -> Result<(BertModel, Tokenizer)> {
    let device = candle_examples::device(false)?;
    let repo = Repo::with_revision("sentence-transformers/all-MiniLM-L6-v2".to_string(), RepoType::Model, "refs/pr/21".to_string());
    let (config_filename, tokenizer_filename, weights_filename) = {
        let api = Api::new()?.repo(repo);
        let config = api.get("config.json")?;
        let tokenizer = api.get("tokenizer.json")?;
        let weights = api.get("model.safetensors")?;
        (config, tokenizer, weights)
    };
    let config = std::fs::read_to_string(config_filename)?;
    let config: Config = serde_json::from_str(&config)?;
    let tokenizer = Tokenizer::from_file(tokenizer_filename).map_err(E::msg)?;
    let vb = unsafe { VarBuilder::from_mmaped_safetensors(&[weights_filename], DTYPE, &device)? };
    let model = BertModel::load(vb, &config)?;
    Ok((model, tokenizer))
}

pub fn normalize_l2(v: &Tensor) -> Result<Tensor> {
    Ok(v.broadcast_div(&v.sqr()?.sum_keepdim(1)?.sqrt()?)?)
}

fn main() -> Result<()> {
    let sentence_count = 500;
    let (model, mut tokenizer) = build_model_and_tokenizer()?;
    let device = &model.device;
    let sentences: Vec<&str> = vec!["This is a sample sentence that should be short enough"; sentence_count];
    if let Some(pp) = tokenizer.get_padding_mut() {
        pp.strategy = tokenizers::PaddingStrategy::BatchLongest
    } else {
        let pp = PaddingParams {
            strategy: tokenizers::PaddingStrategy::BatchLongest,
            ..Default::default()
        };
        tokenizer.with_padding(Some(pp));
    }
    let tokens = tokenizer.encode_batch(sentences, true).map_err(E::msg)?;
    let token_ids = tokens
        .iter()
        .map(|tokens| {
            let tokens = tokens.get_ids().to_vec();
            Ok(Tensor::new(tokens.as_slice(), device)?)
        })
        .collect::<Result<Vec<_>>>()?;

    let token_ids = Tensor::stack(&token_ids, 0)?;
    let token_type_ids = token_ids.zeros_like()?;
    println!("running inference on batch {:?}", token_ids.shape());
    let start = std::time::Instant::now();
    let embeddings = model.forward(&token_ids, &token_type_ids)?;
    println!("model forward timing {:?}", start.elapsed());
    let (_n_sentence, n_tokens, _hidden_size) = embeddings.dims3()?;
    let embeddings = (embeddings.sum(1)? / (n_tokens as f64))?;
    let embeddings = normalize_l2(&embeddings)?;
    println!("pooled embeddings {:?}", embeddings.shape());
    Ok(())
}
```

Note: you might need to increase the value of `sentence_count` to reproduce, for me the cutoff is somewhere between 80 and 100 (i.e  80 ✅  but 100 🔴 ).

### Output:
```
running inference on batch [100, 12]
[1]    52157 abort      cargo run --example bert-metal --features metal
```

For reference, when using the python sentence_transformers library, I am able to embed a much larger number of 
sentences at once without the program getting aborted (sample code below).
```
import torch
from sentence_transformers import SentenceTransformer
import time

if __name__ == "__main__":
    sentences = ["This is a sample sentence that should be short enough"] * 1000
    device = torch.device("mps")
    model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2", device=device)
    start = time.monotonic()
    embeddings = model.encode(sentences, batch_size=len(sentences), device=device)
    done = time.monotonic()
    print(f"embedding (including tokenization) {done - start} seconds")
```
sentence-transformers = "2.2.2"

**Comment by FL33TW00D** [comment]
I can reproduce and have isolated to the [following line](https://github.com/huggingface/candle/blob/main/candle-metal-kernels/src/lib.rs#L1407).

Debugging produces the following:
`-[MTLDebugComputeCommandEncoder setBytes:length:attributeStride:atIndex:]:400: failed assertion length(38400) must be <= 4096.'`
Metal limits this to 4kb: https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443159-setbytes

The matmul occuring is of the following dimensions:
`[100, 12, 12, 32] @ [100, 12, 32, 12]`

The batch dimension becomes 1200, and we allocate a vector [here](https://github.com/huggingface/candle/blob/main/candle-metal-kernels/src/lib.rs#L1400) to write (1200 * 4 * 8 == 38400).

Successfully for me if you change that line to `u32` (despite still going over the documented limit).

**Comment by FL33TW00D** [comment]
I can reproduce and have isolated to this line

> I can reproduce and have isolated to the [following line](https://github.com/huggingface/candle/blob/main/candle-metal-kernels/src/lib.rs#L1407).
> 
> Debugging produces the following: `-[MTLDebugComputeCommandEncoder setBytes:length:attributeStride:atIndex:]:400: failed assertion length(38400) must be <= 4096.'` Metal limits this to 4kb: https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443159-setbytes
> 
> The matmul occuring is of the following dimensions: `[100, 12, 12, 32] @ [100, 12, 32, 12]`
> 
> The batch dimension becomes 1200, and we allocate a vector [here](https://github.com/huggingface/candle/blob/main/candle-metal-kernels/src/lib.rs#L1400) to write (1200 * 4 * 8 == 38400).
> 
> Runs successfully for me if you change that line to `u32`.

**Comment by aorticweb** [comment]
I was able to reproduce, @FL33TW00D's fix, and that increased the maximum number of sentence before the system aborting the program to 170, but I am still confused why the program get's aborted at all, when using python sentence transformers, even if embedding 10000 sentence, the GPU usage spikes to 100%, the embedding take a long time to generate but the program does not get aborted

**Comment by FL33TW00D** [comment]
> I was able to reproduce, @FL33TW00D's fix, and that increased the maximum number of sentence before the system aborting the program to 170, but I am still confused why the program get's aborted at all, when using python sentence transformers, even if embedding 10000 sentence, the GPU usage spikes to 100%, the embedding take a long time to generate but the program does not get aborted

A deeper fix is required, I am working on it today.

**Comment by FL33TW00D** [comment]
### Crux of the problem

[encoder.setBytes](https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443159-setbytes) only supports 4096 bytes of data. We call it on this line with the **byte strides** required [here in MFA](https://github.com/philipturner/metal-flash-attention/blob/2623b778a7a6279ca1ec70f5b4ee825ccd65ffb6/Sources/GEMM.metal#L269). These are used by the GEMM for indexing.

Problem is: it requires 4 strides per batch dim (1 for A matrix, 1 for B matrix, 1 for output matrix, 1 for bias matrix), and these are [supposed to be U64](https://github.com/philipturner/metal-flash-attention/blob/main/Tests/Operations/GEMM.swift#L237) (8 bytes).

Therefore, we can determine the batch limit in MFA is `4096 / (4 * 8)` or 128.

### Resolutions

Workarounds
1. We don't currently support bias on `call_gemm`, so we can remove the last byte stride, which gives us a batch limit of 170.
2. Change the data type to `u32` as above, this gives us a batch limit of 256.

We could then split any batch above 256 into multiple calls.

Or - we can patch MFA to compute the byte strides inside the inner loop, which may be detrimental to performance.

**Comment by ivarflakstad** [comment]
I'm not 100% certain, but I think we could just use a buffer instead of `setBytes` 🤔
